src/proxyvisor.js

```diff
@@ -0,0 +1,1012 @@
+					.catch(() =>
+						Promise.using(
+							proxyvisor.docker.imageRootDirMounted(app.image),
+							rootDir => getTarArchive(rootDir + '/assets', dest),
+						),
+					)
+					.then(() => res.sendFile(dest));
+			})
+			.catch(function(err) {
+				log.error(`Error on ${req.method} ${url.parse(req.url).pathname}`, err);
+				return res.status(503).send(err?.message || err || 'Unknown error');
+			}),
+	);
+
+	router.get('/v1/dependent-apps', (req, res) =>
+		db
+			.models('dependentApp')
+			.select()
+			.map(app => ({
+				id: parseInt(app.appId, 10),
+				commit: app.commit,
+				name: app.name,
+				config: JSON.parse(app.config ?? '{}'),
+			}))
+			.then(apps => res.json(apps))
+			.catch(function(err) {
+				log.error(`Error on ${req.method} ${url.parse(req.url).pathname}`, err);
+				return res.status(503).send(err?.message || err || 'Unknown error');
+			}),
+	);
+
+	return router;
+};
+
+export default class Proxyvisor {
+	constructor({ config, logger, db, docker, images, applications }) {
+		this.bindToAPI = this.bindToAPI.bind(this);
+		this.executeStepAction = this.executeStepAction.bind(this);
+		this.getCurrentStates = this.getCurrentStates.bind(this);
+		this.normaliseDependentAppForDB = this.normaliseDependentAppForDB.bind(
+			this,
+		);
+		this.setTargetInTransaction = this.setTargetInTransaction.bind(this);
+		this.getTarget = this.getTarget.bind(this);
+		this._getHookStep = this._getHookStep.bind(this);
+		this.nextStepsForDependentApp = this.nextStepsForDependentApp.bind(this);
+		this.getRequiredSteps = this.getRequiredSteps.bind(this);
+		this.getHookEndpoint = this.getHookEndpoint.bind(this);
+		this.sendUpdate = this.sendUpdate.bind(this);
+		this.sendDeleteHook = this.sendDeleteHook.bind(this);
+		this.sendUpdates = this.sendUpdates.bind(this);
+		this.config = config;
+		this.logger = logger;
+		this.db = db;
+		this.docker = docker;
+		this.images = images;
+		this.applications = applications;
+		this.acknowledgedState = {};
+		this.lastRequestForDevice = {};
+		this.router = createProxyvisorRouter(this);
+		this.actionExecutors = {
+			updateDependentTargets: step => {
+				return this.config
+					.getMany(['currentApiKey', 'apiTimeout'])
+					.then(({ currentApiKey, apiTimeout }) => {
+						// - take each of the step.devices and update dependentDevice with it (targetCommit, targetEnvironment, targetConfig)
+						// - if update returns 0, then use APIBinder to fetch the device, then store it to the db
+						// - set markedForDeletion: true for devices that are not in the step.devices list
+						// - update dependentApp with step.app
+						return Promise.map(step.devices, device => {
+							const { uuid } = device;
+							// Only consider one app per dependent device for now
+							const appId = _(device.apps)
+								.keys()
+								.head();
+							if (appId == null) {
+								throw new Error(
+									'Could not find an app for the dependent device',
+								);
+							}
```

Comment:
I had to add this to make typescript happy, I assume it never happens but at least if it does we now get a slightly better error message